### PR TITLE
HEEDLS-500 - implement learning log page for delegate progress

### PR DESCRIPTION
**Pull request by DanBloxham-sw**
### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-500

### Description
Implemented new page for the delegate progress learning log. Cleaned up a couple of unrelated pages as well with the formatter.
Slightly modified the SearchSortFilterPaginate JS to work with the table here. Required surprisingly few modifications, mostly making sure that unnecessary logic is not executed when pagination is set to false.

### Screenshots
Desktop:
Learning log desktop

Tablet:

Mobile:

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [ ] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data/DataServices/ProgressDataService.cs [hunk 1]
```diff
@@ -227,5 +229,31 @@ public void UnlockProgress(int progressId)
...
+                    UNION ALL SELECT 
+                            aa.[Date] AS [When],
+                            NULL AS Duration,
+                            sec.SectionName AS AssessmentTaken,
+                            aa.Score AS AssessmentScore,
+                            aa.[Status] AS AssessmentStatus
+                        FROM AssessAttempts AS aa 
+                        INNER JOIN dbo.Customisations AS cu ON cu.CustomisationID = aa.CustomisationID
+                        LEFT JOIN Sections AS sec ON sec.ApplicationID = cu.ApplicationID AND sec.SectionNumber = aa.SectionNumber
+                        WHERE aa.ProgressID = @progressId",
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data/DataServices/ProgressDataService.cs:254]: Formatting appears a little strange in this SQL query are there a mix of tabs and spaces?

File: DigitalLearningSolutions.Web.Tests/ViewModels/TrackingSystem/Delegates/DelegateProgress/LearningLogEntryViewModelTests.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    using DigitalLearningSolutions.Web.ViewModels.TrackingSystem.Delegates.DelegateProgress;
+    using FizzWare.NBuilder;
+    using FluentAssertions;
+    using NUnit.Framework;
+
+    public class LearningLogEntryViewModelTests
+    {
+        [TestCase(1, "1")]
+        [TestCase(null, "N/A")]
+
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web.Tests/ViewModels/TrackingSystem/Delegates/DelegateProgress/LearningLogEntryViewModelTests.cs:13]: All tests in this file seem to have this extra newline?

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateProgress/LearningLog.cshtml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+@{
+  ViewData["Title"] = "Delegate learning log";
+  var routeParamsForBackLink = new Dictionary<string, string?> {
+    {
+      "accessedVia", Model.AccessedVia.Name
+    }, {
+      "progressId", Model.ProgressId.ToString()
+    },
+  };
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateProgress/LearningLog.cshtml:15]: I assume this is something the auto formatter is doing?

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
A few things to tidy up, but looks fine overall

File: DigitalLearningSolutions.Web.Tests/ViewModels/TrackingSystem/Delegates/DelegateProgress/LearningLogEntryViewModelTests.cs (hunk 2 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Web.Tests/ViewModels/TrackingSystem/Delegates/DelegateProgress/LearningLogEntryViewModelTests.cs]: No idea how I missed that. Should be sorted now.

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateProgress/LearningLog.cshtml (hunk 3 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateProgress/LearningLog.cshtml:15]: Afraid so. I'm not sure this page will get any further changes to it, so could manually reformat it to something nicer?

Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateProgress/LearningLog.cshtml:15]: Probably not worth it, easier to just leave it how the auto format will do it and cause less hassle in the future

File: DigitalLearningSolutions.Data/DataServices/ProgressDataService.cs (hunk 1 above)
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data/DataServices/ProgressDataService.cs]: Yeah I think it's better with the UNION ALL on a separate line, almost would be nice to have a way to do it like
```
    SELECT ...
UNION
    SELECT ...
```
but I think that looks even stranger.

**Review by AlexJacksonDS — APPROVED** [review]

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateProgress/LearningLogViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        public int ProgressId { get; set; }
+        public int CustomisationId { get; set; }
+        public string CourseName { get; set; }
+        public int DelegateId { get; set; }
+        public string? DelegateFirstName { get; set; }
+        public string DelegateLastName { get; set; }
+        public string? DelegateEmail { get; set; }
+        public IEnumerable<LearningLogEntryViewModel> Entries { get; set; }
+
+        public string DelegateFullName =>
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateProgress/LearningLogViewModel.cs:43]: This hasn't been added to the ticket, but we should be showing names as `LastName, FirstName` anywhere we sort by them.

File: DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/searchSortFilterAndPaginate.ts [hunk 5]
```diff
@@ -219,6 +224,10 @@ export class SearchSortFilterAndPaginate {
   }
 
   private updatePageNumber(pageNumber: number, searchableData: ISearchableData): void {
+    if (this.paginationEnabled === false) {
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/searchSortFilterAndPaginate.ts:227]: Could we include this condition in the method name?
`updatePageNumberIfPaginated`?

File: DigitalLearningSolutions.Data/Models/LearningLogEntry.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
+﻿namespace DigitalLearningSolutions.Data.Models
+{
+    using System;
+
+    public class LearningLogEntry : BaseSearchableItem
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/Models/LearningLogEntry.cs:5]: Why does this need to be a SearchableItem?

**Review by SteveJacksonSoft — COMMENTED** [review]
Haven't looked at tests yet, but this is all I got on everything else. Looks nice!

File: DigitalLearningSolutions.Data/Models/LearningLogEntry.cs (hunk 6 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Data/Models/LearningLogEntry.cs:5]: The view model's SortItems method requires the object passed in to inherit BaseSearchableItem.

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateProgress/LearningLogViewModel.cs (hunk 4 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateProgress/LearningLogViewModel.cs:43]: I had changed this to use `LastName, FirstName` but then realised we don't sort by the name here. I've changed it back, but left the new string display method in the helper. Do we want to get rid of that as well or leave it for whenever we fix the rest of the names?

Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateProgress/LearningLogViewModel.cs:43]: Hmm, I think we really want two methods - one for sortable names and one for others.
If we don't sort by this name it should be the standard `Forename Surname`.
Feel free to use a helper method, but could you make it clear whether it's for sortable names or non-sortable names? Maybe there's a better adjective to use than that.

File: DigitalLearningSolutions.Web/Helpers/DisplayStringHelper.cs [hunk 7]
```diff
@@ -34,19 +34,24 @@ public static string GenerateSizeDisplayString(long byteCount)
...
+            return Math.Sign(byteCount) * number + Units[place];
         }
 
         public static string? ConvertNumberToMonthsString(int numberOfMonths)
         {
             return numberOfMonths == 0 ? null : $"{numberOfMonths} month{GetPluralitySuffix(numberOfMonths)}";
         }
 
-        public static string GetDelegateNameString(string? firstName, string lastName)
+        public static string GetFullNameForDisplay(string? firstName, string lastName)
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web/Helpers/DisplayStringHelper.cs:45]: Hmm, these names make it seem like we're always displaying it as `Firstname Lastname` but when we sort we're going to sort by Lastname first (but still display `Firstname Lastname`.
Can we make it more clear that the other method gives the full name to display for items which can be sorted by name, and that this method gives the full name to display for items that aren't being sorted by name?

File: DigitalLearningSolutions.Web.Tests/Helpers/DisplayStringHelperTests.cs [hunk 8]
```diff
@@ -134,25 +134,54 @@ public void ConvertNumberToMonthsString_should_return_months_for_more_than_one()
...
+            var result = DisplayStringHelper.GetFullNameForDisplay("FirstName", "LastName");
 
             // Then
             result.Should().Be("FirstName LastName");
         }
 
+        [TestCase(null)]
+        [TestCase("")]
+        [TestCase(" ")]
+        public void GetFullNameForSorting_returns_last_name_only_when_first_name_is_not_a_name(string? firstName)
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web.Tests/Helpers/DisplayStringHelperTests.cs:159]: `...when_first_name_is_null_or_whitespace` would be clearer. Is `932fancv90a` a name? 😛 Guess it depends what platform you're on...

**Review by SteveJacksonSoft — COMMENTED** [review]
Reviewed the tests now as well.

File: DigitalLearningSolutions.Web/Helpers/DisplayStringHelper.cs (hunk 7 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Web/Helpers/DisplayStringHelper.cs]: I've renamed them to hopefully make them clearer. If you think they still aren't obvious enough, I might need some suggestions, as I've wracked my brain for direct and clean names and these are the best I can think of.

**Review by SteveJacksonSoft — APPROVED** [review]
Nice, merge away.

**Pull request merged**